File: 2023/4/main.hs
import Data.List (intersect)
import qualified Data.Map as M
import Text.Regex (matchRegex, mkRegex)

parse :: String -> (Int,[Int],[Int])
parse s = do
    let rx = mkRegex "Card +([0-9]+): ([0-9 ]+) \\| ([0-9 ]+)"
    let Just [cardnum,wins,mynums] = matchRegex rx s
    (read cardnum,map read $ words wins,map read $ words mynums)

main :: IO ()
main = do
    inp <- lines <$> readFile "input.txt"
    let cards = map parse inp
    let intersections = map (\(c,w,n) -> (c,length $ w `intersect` n)) cards
    
    -- Part 1
    let points = sum $ map (\(_,l) -> if l == 0 then 0 else 2^(l-1)) intersections
    putStrLn $ "Part 1: " ++ show points

    -- Part 2
    let cardTotals = M.fromList [(c,1) | (c,_) <- intersections]
    let totals = foldl combine cardTotals intersections
    putStrLn $ "Part 2: " ++ show (M.foldl (+) 0 totals)

combine :: M.Map Int Int -> (Int,Int) -> M.Map Int Int
combine totals (c,n) = M.unionWith (+) totals $ M.fromList [(cc, totals M.! c) | cc <- [c+1 .. c+n]]

File: 2023/4/main.py
import re

regex = re.compile("Card +([0-9]+): ([0-9 ]+) \| ([0-9 ]+)")

with open("input.txt") as f:
    data = f.readlines()

matches = map(lambda x: re.match(regex, x), data)
numbers = {int(m.group(1)): (set(filter(lambda x: x != "", m.group(2).split(" "))), 
            set(m.group(3).split(" "))) for m in matches}
intersections = {card: len(x[0].intersection(x[1])) for card,x in numbers.items()}

# Part 1
points = map(lambda x: 2**(x-1) if x else 0, intersections.values())
print(sum(points))

# Part 2
cards = {card_num: 1 for card_num in intersections.keys()}
for card,nums in intersections.items():
    for i in range(nums):
        cards[card+i+1] += cards[card]
print(sum(cards.values()))
